### ISSUE TITLE
watchlist request will be closed abnormally when cacheInterval contains a large amount of watchEvents

### ISSUE THREAD

**Issue by Chaunceyctx** [kubernetes/kubernetes#127553]
### What happened?

I'm conducting an investigation into watchList benchmark in 1.30 Kubernetes cluster. I found that using `watchList` request to make the reflector become synced takes longer than using `list` request. 

Then, I found that the apiserver always prints `Forcing %s watcher close due to unresponsiveness: key.... ` log.  I realized that if `cacheInterval ` containes lots of watchEvents, then cacheWatcher.Process will not be invoked until all watchEvents in `cacheInterval` have been sent. As a result, the `input` channel in `cacheWatcher` may become full, leading to the abnormal closure of the watchlist request.

Therefore, are there any solutions to address this issue? Perhaps we need a new watchList request handler to process watchList requests instead of reusing the original watch request handler's logic. 

If there isn't a similar issue in this problem, we can use the current issue to track it. :)

### What did you expect to happen?

Using watchList request to make the reflector become synced takes approximately the same amount of time as using list request.

### How can we reproduce it (as minimally and precisely as possible)?

1. create 5K Pods
2. use watchlist request to make reflector become synced. At the same time, start many client to update pods
3. watchlist request will be close and apiserver prints `Forcing %s watcher close due to unresponsiveness: key.... ` log

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by xuzhenglun** [comment]
I'm not sure that this has direct connection with `WatchList`, because Reflector only use `WatchList` as a altnative of `List` for now. `WatchList` request will be closed once the special annotation received, and then  another standard `Watch` will be initiated.

**Comment by xuzhenglun** [comment]
does this can reproduced if watchlist is disabled? maybe it's a performance regression introduced by watchlist related changes? 🤔

**Comment by Chaunceyctx** [comment]
> I'm not sure that this has direct connection with `WatchList`, because Reflector only use `WatchList` as a altnative of `List` for now. `WatchList` request will be closed once the special annotation received, and then another standard `Watch` will be initiated.
> 
> see more, `r.watchlist` as an altnative of `r.list`, and finally `r.watchWithResync` will be called anyway.
> 
> https://github.com/kubernetes/kubernetes/blob/65fc1bb463c85a4c85e619bf7acac9503e23a253/staging/src/k8s.io/client-go/tools/cache/reflector.go#L340-L370

@xuzhenglun 
Yes, this issue will also occur in scenario where resources are frequently updated for `watch` request, not just for `watchlist` request. 

However, the impact of this issue on `watch` requests and `watchlist` requests is different. For `watch` request, after it is closed, `reflector` will restart the `watch` request with a more recent `lastSyncResourceVersion`, which can lead to a reduction in the number of `watchEvents` existing in the `cacheInterval`. This might prevent the `watch` request from being closed. 

On the other hand, for `watchlist` request, after it is closed, `reflector` can only restart the `watchlist` request with an older lastSyncResourceVersion or even ""(empty resourceVersion), resulting in the number of `watchEvents` existing in the `cacheInterval` remaining constant. This leads to the `watchlist` request being continuously closed and retried.

**Comment by xuzhenglun** [comment]
AFAIK, `lastSyncResourceVersion` would be updated once `list` and `watchlist` success or received a event from `watch`. And later once `watch` is closed, then`list` or `watchlist` will be re-initiated which called `relist/rewatch`. In this phase, both `list` and `watchlist` will try to use `lastSyncResourceVersion` to fetch full date.  In this moment if the resourceVersion is expired, `relist/rewatch` will fetch the latest data by using `""` empty resourceVersion.  

So I can not tell any difference of lastSyncResourceVersion between `watchlist` and `list` scenario.  I don't understand why lastSyncResourceVersion is more recent from `watch`. Do you mean that `watchlist` takes more time, so the time window between `watchlist` and `watch` is wider than `watch` and `list`? After all, if this time window lasts more wide, the RV used when executing the `watch` will be relatively older.

**Comment by Chaunceyctx** [comment]
> AFAIK, `lastSyncResourceVersion` would be updated once `list` and `watchlist` success or received a event from `watch`. And later once `watch` is closed, then`list` or `watchlist` will be re-initiated which called `relist/rewatch`. In this phase, both `list` and `watchlist` will try to use `lastSyncResourceVersion` to fetch full date. In this moment if the resourceVersion is expired, `relist/rewatch` will fetch the latest data by using `""` empty resourceVersion.
> 
> So I can not tell any difference of lastSyncResourceVersion between `watchlist` and `list` scenario. I don't understand why lastSyncResourceVersion is more recent from `watch`. Do you mean that `watchlist` takes more time, so the time window between `watchlist` and `watch` is wider than `watch` and `list`? After all, if this time window lasts more wide, the RV used when executing the `watch` will be relatively older.

I will draw flowchart to clarify this issue :)

**Comment by Chaunceyctx** [comment]
```
                                    client                                     server
                                        |                                      |
                                        |                                      |
                                        |                                      |
                                        |                                      |
client starts a watchlist request     ->|                                      |
about resource A with rv=x              |                                      |
                                        |                                      |<- server handles this request and creates a cacheInterval
                                        |                                      |   containing so many watch events
                                        |                                      |
                                        |                                      |
                                        |                                      |<- server invokes processInterval method to send watch events
                                        |                                      |   existing in cacheInterval.At the same time, there are so many 
                                        |                                      |   update/create operation about this resource. And it will lead
                                        |                                      |   to the input channel in cacheWatcher of this watchlist request
                                        |                                      |   are full of watchEvents. 
                                        |                                      |
                                        |                                      |
                                        |                                      |<- server continues to push watch events into the input channel
                                        |                                      |   it is possible that the watchlist request will be closed,
                                        |                                      |   and the io.EOF error will be returned to the streamWatcher 
                                        |                                      |   on the client side.
                                        |                                      |
When the client receives the io.EOF   ->|                                      |
error and has not received the initial  |                                      |
bookmark event, it will restart the     |                                      |
watchlist request with rv=x             |                                      |
                                        |                                      |
                                        |                                      |
                                        |                                      |
                                        |                                      |<- server handles this request and still create a cacheInterval 
                                        |                                      | containing so many watch events. Finally, this watchlist 
                                        |                                      | request will be closed again.
```

@xuzhenglun there is a flowchart which can better illustrate this issue

**Comment by xuzhenglun** [comment]
@Chaunceyctx 
Really thanks for your patient flowchat, it's really helpful. Here is my understanding, correct me if i'm wrond.

Compared to traditional `list` and then `watch`, `watchlist` has more events need to sent from `cacheInterval`. it because in traditional `list&watch` machiasm, `cacheInterval` only need to send events incremented since the version returned by list, but `watchlist` need to send all events. so it will take more time to do it, and more likely cause `input channel` full, and finally force close the watchlist request. 

/cc @p0lyn0mial 
1. maybe we can try to add chanel buffer size to reduce the chance of this problems? 
2. in current reflector impliement, watchlist and watch are seperated phases anyway. so maybe we can intro an new option such as `initialEventOnly` which dedicates current request don't need following events. And once initial events are all sent, server could close the requst. so we can drop all events from input channel to prevents this problem. 

And I still have another few questions:
1. for the client side, there is a fallback machiasm which make sure `watchlist` should be able to fallback to traditional `list`, but the machiasm is not always working, right? the reflector just keeping do `watchlist` request.  I'm not sure that does this fallback mechiasm fails forever, or just keep a little while? In current reflector impliment, reflector will re-request watchlist if `watchlist` request closed after 1 second but not received the initial bookmark, and this may cause a deadlock just like you said. But this deadlock should just keep a little while, and can be resolved automatically once a `410 gone / expired` error, and reflector will fallback to tradional `list` AFAIK.

**Comment by Chaunceyctx** [comment]
> @Chaunceyctx Really thanks for your patient flowchat, it's really helpful. Here is my understanding, correct me if i'm wrond.
> 
> Compared to traditional `list` and then `watch`, `watchlist` has more events need to sent from `cacheInterval`. it because in traditional `list&watch` machiasm, `cacheInterval` only need to send events incremented since the version returned by list, but `watchlist` need to send all events. so it will take more time to do it, and more likely cause `input channel` full, and finally force close the watchlist request.
> 
> /cc @p0lyn0mial
> 
> 1. maybe we can try to add chanel buffer size to reduce the chance of this problems?
> 2. in current reflector impliement, watchlist and watch are seperated phases anyway. so maybe we can intro an new option such as `initialEventOnly` which dedicates current request don't need following events. And once initial events are all sent, server could close the requst. so we can drop all events from input channel to prevents this problem.
> 
> And I still have another few questions:
> 
> 1. for the client side, there is a fallback machiasm which make sure `watchlist` should be able to fallback to traditional `list`, but the machiasm is not always working, right? the reflector just keeping do `watchlist` request.  I'm not sure that does this fallback mechiasm fails forever, or just keep a little while? In current reflector impliment, reflector will re-request watchlist if `watchlist` request closed after 1 second but not received the initial bookmark, and this may cause a deadlock just like you said. But this deadlock should just keep a little while, and can be resolved automatically once a `410 gone / expired` error, and reflector will fallback to tradional `list` AFAIK.

IMO,
option1: the results of adjusting the input channel size sounds unpredictable. Perhaps a large number of testing experiments are needed to support the adjustment results.
option2: yes, we may need to differentiate between watchlist and watch at least, and we should not re-use the watch handler to process watchlist request.

If the `watchHandler` returns `err != nil && err != 410 gone / expired`, it will automatically fallback to the list. However, if the `watchHandler` returns `err=nil`(apiserver always closes `watchlist` request actively), it will cause the `watchlist` request to keep retrying indefinitely and never exit.

**Comment by xuzhenglun** [comment]
> If the watchHandler returns err != nil && err != 410 gone / expired, it will automatically fallback to the list. However, if the watchHandler returns err=nil(apiserver always closes watchlist request actively), it will cause the watchlist request to keep retrying indefinitely and never exit.

Yeah, reflector will keep retry `watchlist` only if an error returns. What I'm curious is that if the resource is keep updating frequently, the rv=x should be expired soon.  because the upper bound of watchcache size is `100 * 1024`.

>  the results of adjusting the input channel size sounds unpredictable. 
yes,  you are right,  maybe we really need to do some perforance perf to guide us how to optomize this.  here is just my throught, maybe the size could be estimated by the count of initial events need to send. 

> we may need to differentiate between watchlist and watch at least, and we should not re-use the watch handler to process watchlist request.
I'm not sure what will happen in future about `watchlist`. In current impliment, `watchlist` is basicly just an altnative of `list`. reflector will close the `watchlist` and then do another `watch`. I think it is possible to merge those two phases and this could reduce an api call. But I'm not sure is this in the current story of `watchlist`.
